packages/core/src/di/r3_injector.ts

```diff
@@ -95,35 +101,203 @@ export function createInjector(
+  let def = getInjectorDef(container);
+
+  // If that's not present, then attempt to read ngModule from the InjectorDefTypeWithProviders.
+  const ngModule =
+      (def == null) && (container as InjectorTypeWithProviders<any>).ngModule || undefined;
+
+  // Determine the InjectorType. In the case where `defOrWrappedDef` is an `InjectorType`,
+  // then this is easy. In the case of an InjectorDefTypeWithProviders, then the definition type
+  // is the `ngModule`.
+  const defType: InjectorType<any> =
+      (ngModule === undefined) ? (container as InjectorType<any>) : ngModule;
+
+  // Check for circular dependencies.
+  if (ngDevMode && parents.indexOf(defType) !== -1) {
+    const defName = stringify(defType);
+    const path = parents.map(stringify);
+    throwCyclicDependencyError(defName, path);
+  }
+
+  // Check for multiple imports of the same module
+  const isDuplicate = dedup.has(defType);
+
+  // Finally, if defOrWrappedType was an `InjectorDefTypeWithProviders`, then the actual
+  // `InjectorDef` is on its `ngModule`.
+  if (ngModule !== undefined) {
+    def = getInjectorDef(ngModule);
+  }
+
+  // If no definition was found, it might be from exports. Remove it.
+  if (def == null) {
+    return false;
+  }
+
+  // Add providers in the same way that @NgModule resolution did:
+
+  // First, include providers from any imports.
+  if (def.imports != null && !isDuplicate) {
+    // Before processing defType's imports, add it to the set of parents. This way, if it ends
+    // up deeply importing itself, this can be detected.
+    ngDevMode && parents.push(defType);
+    // Add it to the set of dedups. This way we can detect multiple imports of the same module
+    dedup.add(defType);
+
+    let importTypesWithProviders: (InjectorTypeWithProviders<any>[])|undefined;
+    try {
+      deepForEach(def.imports, imported => {
+        if (walkProviderTree(imported, providersOut, parents, dedup)) {
+          if (importTypesWithProviders === undefined) importTypesWithProviders = [];
+          // If the processed import is an injector type with providers, we store it in the
+          // list of import types with providers, so that we can process those afterwards.
+          importTypesWithProviders.push(imported);
+        }
+      });
+    } finally {
+      // Remove it from the parents set when finished.
+      ngDevMode && parents.pop();
+    }
+
+    // Imports which are declared with providers (TypeWithProviders) need to be processed
+    // after all imported modules are processed. This is similar to how View Engine
+    // processes/merges module imports in the metadata resolver. See: FW-1349.
+    if (importTypesWithProviders !== undefined) {
+      for (let i = 0; i < importTypesWithProviders.length; i++) {
+        const {ngModule, providers} = importTypesWithProviders[i];
+        deepForEach(providers!, provider => {
+          validateProvider(provider, providers || EMPTY_ARRAY, ngModule);
```

Comment:
Should we guard it with the `ngDevMode` flag?
```suggestion
          ngDevMode && validateProvider(provider, providers || EMPTY_ARRAY, ngModule);
```